apply php 7.2 to travis

samsonasik opened the pull request.

samsonasik committed:
apply php 7.2 to travis

samsonasik commented:
Got following errrors under php 7.2: 
```
There were 6 errors:

1) ZendTest\Session\Config\SessionConfigTest::testSavePathCanBeNonDirectoryWhenSaveHandlerNotFiles

Zend\Session\Exception\InvalidArgumentException: Invalid save handler specified: ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:148

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:68

2) ZendTest\Session\Config\SessionConfigTest::testSaveHandlerIsMutable

ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:97

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:147

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:912

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:105

3) ZendTest\Session\Config\SessionConfigTest::testSaveHandlerAltersIniSetting

ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:97

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:147

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:912

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:111

4) ZendTest\Session\Config\SessionConfigTest::testSetOptionSetsIniSetting with data set #2 ('save_handler', 'getOption', 'user')

ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:97

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:972

5) ZendTest\Session\Config\SessionConfigTest::testSetOptionsTranslatesUnderscoreSeparatedKeys with data set #2 ('save_handler', 'getOption', 'user')

ini_set(): Cannot set 'user' save handler by ini_set() or session_module_name()

/home/travis/build/zendframework/zend-session/src/Config/SessionConfig.php:97

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:147

/home/travis/build/zendframework/zend-session/src/Config/StandardConfig.php:116

/home/travis/build/zendframework/zend-session/test/Config/SessionConfigTest.php:1005

6) ZendTest\Session\SaveHandler\MongoDBTest::testReadDestroysExpiredSession

ini_set(): Headers already sent. You cannot change the session module's ini settings at this time

/home/travis/build/zendframework/zend-session/test/SaveHandler/MongoDBTest.php:97
```

michalbundyra commented on .travis.yml:
```diff
@@ -51,17 +51,17 @@ matrix:
         - DEPS=latest
   allow_failures:
-    - php: nightly
+    - php: 7.2
```
Not sure if I am right, but somewhere, someone secretly decided to not allow failures on PHP 7.2 any more ;-)
@froschdesign Is it right? :smile:

froschdesign replied:
@samsonasik and @webimpress 
November 30, 2017 is around the corner therefore it make no sense! :smiley:

samsonasik replied:
@webimpress @froschdesign I've removed allow_failures->php 7.2, do you think we need to keep the "nightly"?

michalbundyra replied:
@samsonasik I wouldn't do that as long as we don't have PHP 7.3-alpha yet.

samsonasik committed:
removed allow_failures ->  php 7.2

samsonasik commented:
So, any suggestion to get the error gone in php 7.2 ?

michalbundyra commented:
@samsonasik I'd suggest to try [`session_set_save_handler()`](http://php.net/manual/en/function.session-set-save-handler.php) method, because error message says:

```
Cannot set 'user' save handler by ini_set() or session_module_name()
```

michalbundyra commented:
@samsonasik Hm... oh, no... sorry. No idea for now. This is not gonna work.

remicollet commented:
Same issue from Fedora QA
https://apps.fedoraproject.org/koschei/package/php-zendframework-zend-session?collection=f28

weierophinney committed:
Merge pull request #92 from samsonasik/patch-1

apply php 7.2 to travis

weierophinney commented:
This is a case where the tests are incorrect; the value "user" was never correct, but previous versions of PHP did not raise an error.

I'll update to provide a mock `SessionHandlerInterface` implementation.

weierophinney committed:
Ensure save handler configuration works with PHP 7.2

This patch takes the following approach:

- If "files" is provided in _any_ PHP version, it continues to use
  `ini_set`.
- If "user" is provided in PHP versions prior to 7.2, it continues to
  use `ini_set`.
- Otherwise, it checks to see if the handler provided is a class or
  instance that implements PHP's `SessionHandlerInterface`; if so, it
  uses `session_set_save_handler()` with that instance (instantiating it
  if it is a class name).
- It no longer uses `ini_get()` to retrieve the save handler
  configuration _unless_ it has not been memoized in the instance.

This approach works cross-platform.

weierophinney committed:
ext/mongodb installation in Travis-CI

- PHP 7+ environments have it by default, but it's out-of-date
- PHP 5 environments have ext/mongo but not ext/mongodb, so we need to
  install it.

`pecl upgrade` will install if it's not present, and upgrade otherwise.
Once installed, we need to enable it if it is not already.

michalbundyra commented on src/Config/SessionConfig.php:
```diff
@@ -126,11 +136,31 @@ public function getStorageOption($storageOption)
+     */
+    public function setSaveHandler($phpSaveHandler)
+    {
+        // error_log(sprintf('In %s with %s', __METHOD__, var_export($phpSaveHandler, true)));
```
Do we need this comment above?

weierophinney replied:
I'll remove; thanks!

michalbundyra commented on src/SessionManager.php:
```diff
@@ -126,13 +127,23 @@ public function start($preserveStorage = false)
+            // ArrayUtils::merge parameter
+            if ($oldSessionData instanceof Storage\StorageInterface) {
+                $oldSessionData = $oldSessionData->toArray();
+            } elseif ($oldSessionData instanceof \Traversable) {
```
`Traversable` is imported

michalbundyra commented on src/SessionManager.php:
```diff
@@ -126,13 +127,23 @@ public function start($preserveStorage = false)
+            if ($oldSessionData instanceof Storage\StorageInterface) {
+                $oldSessionData = $oldSessionData->toArray();
+            } elseif ($oldSessionData instanceof \Traversable) {
+                $oldSessionData = iterator_to_array($oldSessionData);
```
I'm not sure if it is sufficient - because it is not recursive.

michalbundyra commented on src/SessionManager.php:
```diff
@@ -126,13 +127,23 @@ public function start($preserveStorage = false)
-        if ($oldSessionData instanceof \Traversable
-            || (! empty($oldSessionData) && is_array($oldSessionData))
-        ) {
+        if ($oldSessionData instanceof Traversable) {
```
As noted by @Xerkus in separate PR this is no longer needed

michalbundyra commented on src/SessionManager.php:
```diff
@@ -126,13 +127,23 @@ public function start($preserveStorage = false)
+            $oldSessionData = $oldSessionData->toArray();
+        }
+
+        if (! empty($oldSessionData) && is_array($oldSessionData)) {
```
`! empty` is useless

weierophinney committed:
Ensure headers are not sent for test case

Test case fails on PHP 7.2 due to headers being sent; run in separate
process.

weierophinney commented:
Single test failure is on 5.6-locked, where the MongoDb SaveHandler implementation's garbage collection test reliably fails; passes on all other environments, however, so I'm going to take it as a pass.

weierophinney commented:
Thanks, @samsonasik, for getting this started!
